bee-plugin/src/handler.rs

```diff
@@ -0,0 +1,218 @@
+                }
+            )*
+        }
+    }};
+}
+
+/// A handler for a plugin.
+pub(crate) struct PluginHandler {
+    /// The identifier of the plugin.
+    plugin_id: PluginId,
+    /// The name of the plugin.
+    name: String,
+    /// Shutdown for every `PluginStreamer` used by the plugin.
+    shutdowns: HashMap<EventId, Sender<()>>,
+    /// The OS process running the plugin.
+    process: Child,
+    /// The gRPC client.
+    client: PluginClient<Channel>,
+    /// The task handling stdio redirection.
+    stdio_task: JoinHandle<Result<(), std::io::Error>>,
+}
+
+impl PluginHandler {
+    /// Creates a new plugin handler from a process running the plugin logic.
+    pub(crate) async fn new(
+        plugin_id: PluginId,
+        mut command: Command,
+        bus: &EventBus<'static, UniqueId>,
+    ) -> Result<Self, PluginError> {
+        command.kill_on_drop(true).stdout(Stdio::piped()).stderr(Stdio::piped());
+
+        log::info!(
+            "spawning command `{:?}` for the plugin with ID `{:?}`",
+            command,
+            plugin_id
+        );
+        let mut process = command.spawn()?;
+
+        // stderr and stdout are guaranteed to be `Some` because we piped them in the command.
+        let stderr = BufReader::new(process.stderr.take().unwrap());
+        let mut stdout = BufReader::new(process.stdout.take().unwrap());
+
+        let mut buf = String::new();
+        stdout.read_line(&mut buf).await?;
+        let handshake_info = HandshakeInfo::parse(&buf)?;
+
+        let name = format!("{}-{}", handshake_info.name, plugin_id.0);
+        let target = format!("plugins::{}", name);
+        let stdio_task = tokio::spawn(async move {
+            let mut stdout_lines = stdout.lines();
+            let mut stderr_lines = stderr.lines();
+
+            loop {
+                tokio::select! {
+                    res = stdout_lines.next_line() => match res? {
+                        Some(line) => {
+                            log::info!(target: &target, "{}", line);
+                        },
+                        None => break,
+                    },
+                    res = stderr_lines.next_line() => match res? {
+                        Some(line) => {
+                            log::warn!(target: &target, "{}", line);
+                        },
+                        None => break,
+                    }
+                }
+            }
+
+            Ok(())
+        });
+
+        let address = format!("http://{}/", handshake_info.address);
+        log::info!("connecting to the `{}` plugin at `{}`", name, address);
+        let client = async {
+            let mut count = 0;
+            loop {
+                match PluginClient::connect(address.clone()).await {
+                    Ok(client) => break Ok(client),
+                    Err(err) => {
+                        log::warn!("connection to the `{}` plugin failed: {}", name, err);
```

Comment:
What's the scenario in which a plugin cannot be reached on the first attempt, but after some time becomes reachable again?

Reply:
This happens if the plugin sends a valid handshake before starting its gRPC server. Because the node will try to connect to the server stated in the handshake immediately.